### ISSUE TITLE
Fedora-38-based management qube causes Salt errors when updating certain templates: `Failed to return clean data... ModuleNotFoundError: No module named 'urllib3.util.ssl_match_hostname'`

### ISSUE THREAD

**Issue by Minimalist73** [QubesOS/qubes-issues#8440]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release
4.1

### Brief summary
Updating Debian-11 based templates/qubes with salt results in a salt error about `urllib3.util.ssl_match_hostname` not being found.

### Steps to reproduce
- Open a dom0 terminal
- Update with the following command:
  - `qubesctl --show-output --skip-dom0 --targets debian-11 state.apply update.qubes-vm`

### Expected behavior
Template/qubes based on debian-11 update without issues

### Actual behavior
Salt return an error:
```
Traceback (most recent call last):
            File "/var/tmp/.root_dd8a91_salt/salt-call", line 27, in <module>
              salt_call()
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/scripts.py", line 438, in salt_call
              import salt.cli.call
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/call.py", line 3, in <module>
              import salt.cli.caller
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/caller.py", line 12, in <module>
              import salt.channel.client
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/channel/client.py", line 13, in <module>
              import salt.crypt
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/crypt.py", line 26, in <module>
              import salt.payload
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/payload.py", line 12, in <module>
              import salt.loader.context
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/loader/__init__.py", line 23, in <module>
              import salt.utils.event
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/event.py", line 67, in <module>
              import salt.ext.tornado.iostream
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/iostream.py", line 42, in <module>
              import urllib3.util.ssl_match_hostname
          ModuleNotFoundError: No module named 'urllib3.util.ssl_match_hostname'
          [ERROR   ] An un-handled exception was caught by Salt's global exception handler:
          ModuleNotFoundError: No module named 'urllib3.util.ssl_match_hostname'
          Traceback (most recent call last):
            File "/var/tmp/.root_dd8a91_salt/salt-call", line 27, in <module>
              salt_call()
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/scripts.py", line 438, in salt_call
              import salt.cli.call
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/call.py", line 3, in <module>
              import salt.cli.caller
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/caller.py", line 12, in <module>
              import salt.channel.client
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/channel/client.py", line 13, in <module>
              import salt.crypt
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/crypt.py", line 26, in <module>
              import salt.payload
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/payload.py", line 12, in <module>
              import salt.loader.context
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/loader/__init__.py", line 23, in <module>
              import salt.utils.event
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/event.py", line 67, in <module>
              import salt.ext.tornado.iostream
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/iostream.py", line 42, in <module>
              import urllib3.util.ssl_match_hostname
          ModuleNotFoundError: No module named 'urllib3.util.ssl_match_hostname'
```

**Comment by unman** [comment]
I'm afraid this report is not particularly helpful, as many people are
updating Debian templates without any such issue.
We need to know what is specific to *this* set up that results in the
error.
The "Steps to reproduce " section is missing information about which
Debian-11 template (full or minimal) is used, and what has been done to
it.
Does the error appear for *all* (and only) Debian-11 templates?
Are there Debian templates which update correctly?
Also information re management_dispvm used (check qvm-prefs) and what
template is used by that qube.
Also, have dom0 and other templates been updated?

**Comment by laoclos** [comment]
I am getting exact the same errors using the Qubes built-in update tools also for the default Whonix-16 (which is also based on Debian?!). The debian is a standalone clone of the default debian-11 template with not adjustments. Thus, no special adjustments on both systems. On another system running Qubes all updates were successful three or four days ago, without any issues.

**Comment by Minimalist73** [comment]
> Does the error appear for *all* (and only) Debian-11 templates? Are there Debian templates which update correctly? Also information re management_dispvm used (check qvm-prefs) and what template is used by that qube. Also, have dom0 and other templates been updated?

I tested both debian-11 and debian-12, only debian-11 seems impacted (and other templates like whonix-16 too).
All templates are fully updated, that's when it started to do this.
The management qube use fedora-38 (fedora is not impacted by this).

This started yesterday after a salt update was installed on all my debian-11 based templates (version 3002.6)
Debian 12 use a different salt version (3006.1-2), it can be the reason why it's not impacted.

**Comment by Minimalist73** [comment]
I tested a few more things with the management qube:
- Based on template debian-11:
  - Running salt version `3002.6+dfsg1-4+deb11u1`
  - Working fine for debian-11 templates and no errors anywhere
- Based on template debian-12:
  - Running salt version `3006.1-2+deb12u1`
  - Working fine for debian-11 templates and no errors anywhere
- Based on template fedora-38:
    - Running salt version `3006.2-1`
    - Debian-11 show errors while debian-12 is working fine (difference in salt version between the template can be the reason)

Downgrading the salt version to `3006.1-2` on the fedora-38 template is fixing my issue for debian-11.

It means that the new salt version from the fedora-38 template (3006.2-1) is the reason why debian-11 based templates are showing this module error.

**Comment by unman** [comment]
@Minimalist
Thank you for you detailed and comprehensive work.
It is extremely helpful, and I wish other users were as capable.

**Comment by Szewcson** [comment]
Have the same issue also for some minimal fedora 38 templates. It starts after update of fedora 38 based management qube.

**Comment by gp397** [comment]
@Minimalist73 
I had the same issue with the whonix-16 templates, confirmed downgrading salt in the management template also resolves that issue.

I have temporarily put a versionlock in for salt in the management template on my system.

**Comment by andrewdavidwong** [comment]
Does this also affect 4.2?

**Comment by andrewdavidwong** [comment]
@marmarek: Many users have independently encountered this bug and asked about it (e.g., on the forum and Reddit).

**Comment by marmarek** [comment]
> Does this also affect 4.2?

In the sense of salt failing - probably yes. But since R4.2's updater doesn't use salt anymore, the updater shouldn't be affected.

**Comment by Solomon1732** [comment]
For now, how should 4.1 users update affected qubes? I understood not using salt can miss important update processes. It's why I use it when I update (specifically the graphic interface). So now I'm stuck between potentially partially updating affected qubes and not updating them at all (at least through the graphic interface).

**Comment by marmarek** [comment]
This looks to be caused by a [patch Fedora applies in their packaging](https://src.fedoraproject.org/rpms/salt/blob/rawhide/f/match_hostname.patch), not an upstream release. There is a salt 3006.2-2 package in our testing repo (https://github.com/QubesOS/updates-status/issues/3967) that doesn't have this patch. And for me, it works for debian-11.
Can somebody else also confirm if installing updates from current-testing in fedora-38 fixes the issue?

**Comment by Solomon1732** [comment]
> This looks to be caused by a [patch Fedora applies in their packaging](https://src.fedoraproject.org/rpms/salt/blob/rawhide/f/match_hostname.patch), not an upstream release.

It's not the first time I remember a Fedora patch broke a package. Note I don't hear about it constantly or even frequently. It does happen from time to time, though.

How can the Fedora package maintainer be notified about it? I haven't the slightest how to do it myself.

**Comment by marmarek** [comment]
> How can the Fedora package maintainer be notified about it?

bugzilla.redhat.com I guess

**Comment by rapenne-s** [comment]
> > Does this also affect 4.2?
> 
> In the sense of salt failing - probably yes. But since R4.2's updater doesn't use salt anymore, the updater shouldn't be affected.

this affects the upgrade from 4.1 to 4.2-RC2

https://forum.qubes-os.org/t/qubes-os-4-2-0-rc2-is-available-for-testing/20635/7

**Comment by Minimalist73** [comment]
> Can somebody else also confirm if installing updates from current-testing in fedora-38 fixes the issue?

I can confirm that salt version `3006.2-2` in current-testing allow debian-11 based templates to upgrade without errors using fedora-38 as a management qube.

**Comment by marmarek** [comment]
> I can confirm that salt version `3006.2-2` in current-testing allow debian-11 based templates to upgrade without errors using fedora-38 as a management qube.

Thanks. I'm pushing this package to "current" then.

**Comment by Solomon1732** [comment]
The update to stable solved the issues I had for both whonix-16 and debian 12 minimal qubes.

**Comment by Szewcson** [comment]
I have similar issue on 4.2 when trying to use salt on debian-12-minimal template. On debian-12 everything work.

**Comment by andrewdavidwong** [comment]
> I have similar issue on 4.2 when trying to use salt on debian-12-minimal template. On debian-12 everything work.

That sounds like it's probably a different bug. Please open a separate issue for that, unless one already exists for it.

**Comment by Minimalist73** [comment]
@marmarek Salt updated to version `3006.3-1` today so the issue is back again since it's replacing Qubes salt package in fedora-38 based templates.

**Comment by marmarek** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=2239817

**Comment by Solomon1732** [comment]
I ran into the same error when I tried to update a clean Fedora-38-minimal template. I downloaded it for unrelated purposes and ran into it by happenstance. It's the same error of "no module urllib3 found."

**Comment by Solomon1732** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=2239817
It's been pushed into the stable repository. I confirm it fixed the breaking I had with the following templates: fedora-38-minimal, debian-12-minimal, whonix-16-ws, whonix-16-gw.

**Comment by DemiMarie** [comment]
Closing as fixed.

**Comment by janbue** [comment]
@marmarek @DemiMarie this issue seems to be active again for the fedora-39-minimal template.

**Comment by DemiMarie** [comment]
> @marmarek @DemiMarie this issue seems to be active again for the fedora-39-minimal template.

Which qube are you trying to manage?  The management qube must be as new as the qube it is managing.

**Comment by janbue** [comment]
> > @marmarek @DemiMarie this issue seems to be active again for the fedora-39-minimal template.
> 
> Which qube are you trying to manage?  The management qube must be as new as the qube it is managing.

Trying to manage fedora-39-mimimal from freshly upgraded and updated dom0

**Comment by Enteee** [comment]
@DemiMarie I can confirm this exact same issue.  

My `default-mgmt-dvm` is using `fedora-39` as template.
I just freshly installed `fedora-39-minimal` and then I ran:

```
$ sudo qubesctl --target fedora-39-minimal --skip-dom0 --show-output state.apply
```

and i got the following:
```
fedora-39-minimal:
      ----------
      _error:
          Failed to return clean data
      retcode:
          1
      stderr:
          /var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/util.py:246: SyntaxWarning: invalid escape sequence '\d'
            """Unescape a string escaped by `re.escape`.
          Traceback (most recent call last):
            File "/var/tmp/.root_dd8a91_salt/salt-call", line 27, in <module>
              salt_call()
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/scripts.py", line 437, in salt_call
              import salt.cli.call
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/call.py", line 3, in <module>
              import salt.cli.caller
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/caller.py", line 12, in <module>
              import salt.channel.client
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/channel/client.py", line 13, in <module>
              import salt.crypt
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/crypt.py", line 26, in <module>
              import salt.payload
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/payload.py", line 12, in <module>
              import salt.loader.context
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/loader/__init__.py", line 15, in <module>
              import salt.config
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/config/__init__.py", line 107, in <module>
              _DFLT_IPC_WBUFFER = int(_gather_buffer_space() * 0.5)
                                      ^^^^^^^^^^^^^^^^^^^^^^
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/config/__init__.py", line 95, in _gather_buffer_space
              import salt.grains.core
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/grains/core.py", line 30, in <module>
              import salt.modules.cmdmod
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/modules/cmdmod.py", line 32, in <module>
              import salt.utils.templates
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/templates.py", line 21, in <module>
              import salt.utils.http
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/http.py", line 27, in <module>
              import salt.ext.tornado.simple_httpclient
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/simple_httpclient.py", line 9, in <module>
              from salt.ext.tornado.http1connection import HTTP1Connection, HTTP1ConnectionParameters
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/http1connection.py", line 31, in <module>
              from salt.ext.tornado import iostream
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/iostream.py", line 41, in <module>
              from salt.ext.tornado.netutil import ssl_wrap_socket, ssl_match_hostname, SSLCertificateError, _client_ssl_defaults, _server_ssl_defaults
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/netutil.py", line 57, in <module>
              import urllib3.util.ssl_match_hostname
          ModuleNotFoundError: No module named 'urllib3'
          [ERROR   ] An un-handled exception was caught by Salt's global exception handler:
          ModuleNotFoundError: No module named 'urllib3'
          Traceback (most recent call last):
            File "/var/tmp/.root_dd8a91_salt/salt-call", line 27, in <module>
              salt_call()
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/scripts.py", line 437, in salt_call
              import salt.cli.call
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/call.py", line 3, in <module>
              import salt.cli.caller
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/cli/caller.py", line 12, in <module>
              import salt.channel.client
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/channel/client.py", line 13, in <module>
              import salt.crypt
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/crypt.py", line 26, in <module>
              import salt.payload
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/payload.py", line 12, in <module>
              import salt.loader.context
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/loader/__init__.py", line 15, in <module>
              import salt.config
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/config/__init__.py", line 107, in <module>
              _DFLT_IPC_WBUFFER = int(_gather_buffer_space() * 0.5)
                                      ^^^^^^^^^^^^^^^^^^^^^^
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/config/__init__.py", line 95, in _gather_buffer_space
              import salt.grains.core
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/grains/core.py", line 30, in <module>
              import salt.modules.cmdmod
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/modules/cmdmod.py", line 32, in <module>
              import salt.utils.templates
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/templates.py", line 21, in <module>
              import salt.utils.http
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/utils/http.py", line 27, in <module>
              import salt.ext.tornado.simple_httpclient
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/simple_httpclient.py", line 9, in <module>
              from salt.ext.tornado.http1connection import HTTP1Connection, HTTP1ConnectionParameters
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/http1connection.py", line 31, in <module>
              from salt.ext.tornado import iostream
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/iostream.py", line 41, in <module>
              from salt.ext.tornado.netutil import ssl_wrap_socket, ssl_match_hostname, SSLCertificateError, _client_ssl_defaults, _server_ssl_defaults
            File "/var/tmp/.root_dd8a91_salt/pyall/salt/ext/tornado/netutil.py", line 57, in <module>
              import urllib3.util.ssl_match_hostname
          ModuleNotFoundError: No module named 'urllib3'
      stdout:
  /usr/lib/python3.12/site-packages/salt/utils/jid.py:19: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    return datetime.datetime.utcnow()
```

**Comment by DemiMarie** [comment]
@Enteee does using a Fedora 40-based management qube solve the problem?

**Comment by marmarek** [comment]
It's about `python3-urllib3` in the target qube (fedora-39-minimal here), not management qube. Looks more like https://github.com/QubesOS/qubes-issues/issues/8806#issuecomment-1952233875, and it's supposedly fixed in [fedora-39-minimal 4.2.0-202403250044](https://github.com/QubesOS/updates-status/issues/4460). Which template version do you have? See also [a note in the minimal template docs](https://www.qubes-os.org/doc/templates/minimal/#fedora):
> To manage fedora-39-minimal templates with salt, you may need to install python3-urllib3 in older versions of the template.

**Comment by Enteee** [comment]
@marmarek i removed my fedora-39-minimal template and and reinstalled the version you sugested (`...0044`). Looks like salt is now working for that template. Problem is only really that I have quite a few users which use cloned templates from an old version of 39-minimal. So i probably have to deploy some dom0 `qvm-run dnf install ...` hackaroo for them to fix this up. I am pretty sure this was not an issue with minimal fedora-38 templates.

But yeah, i guess from a qubes perspective this issue is fixed.
